### ISSUE TITLE
Thanks for your quick response!

### ISSUE THREAD

**Issue by dobbelina** [Isayso/PlaylistEditorTV#12]
Thanks for your quick response, awesome!

Being able to hide columns is good, but my intent with the earlier
request was that if a column is hidden, then it wouldn't write that tag
into the playlist either.
Example:
All columns used:
```
#EXTINF:-1 tvg-name="N/A" tvg-id="N/A" group-title="N/A" tvg-logo="https://i.imgur.com/KacMcVa.png",CNN
http://92.43.140.249/s27/04.m3u8
```
First three columns hidden:
```
#EXTINF:-1 tvg-logo="https://i.imgur.com/KacMcVa.png",CNN
http://92.43.140.249/s27/04.m3u8
```

This is how a typical playlist looks like in kodi, that only uses Logo, Name2 and link columns:
<img width="672" alt="screenshot000" src="https://user-images.githubusercontent.com/46063764/66375021-9ed64500-e9ac-11e9-95f8-584a74f1f7be.png">

Here is a typical playlist found on pastebin where the first 2 columns aren't used:
(Name and Id)
https://pastebin.com/raw/jUG8ytNj
I understand that it would be difficult to program it in a way, that it actually deletes
existing content in a column that has been de-selected.
But it doesn't need to do that.
All it has to do is simply ignore writing tags on those de-selected columns.
That should be more easy to accomplish programmatically, i think.
If you manage to pull that off, a really clean kodi playlist could be created,
and the kodi community would love it!

**Comment by Isayso** [comment]
Thanks, now I understand the problem.
I made a fix to write only visible columns and show only columns which have valid entries in it during file load. 
Do you think this could be a solution? A variable column number is more difficult I think.

**Comment by dobbelina** [comment]
Hi again,

This starts to look really promising!
I really like the fact that if a column is hidden, with content written into it's tags,
That gets deleted also when saved, awesome!

There are a few bugs to iron out though.
1. in attached test.m3u it misses that there is content in the logo column when opened,
so that one gets hidden.
I guess the search routine have to look thru all the cells in the column.

2. Copy row works, but insert row from context menu don't work,(Maybe rename paste row also)

3. Move row up and down arrows don't work when there are hidden columns.

4. "Hidden columns avaliable, continue?" pop-up i would remove when saving.

5. rename all instances of "Strg" to "Ctrl", are you german? ha ha lol! :smile:
[test.m3u.zip](https://github.com/Isayso/PlaylistEditorTV/files/3706394/test.m3u.zip)

https://steamcommunity.com/app/447020/discussions/0/312265782624887781/

If you can fix these issues, it will be amazing, and thanks for doing this!

**Comment by Isayso** [comment]
Thanks for your troubleshooting!
More features more problems and side effects.

1. I found an error, so fixed
2. Thats more work. I used the normal clipboard copy functionality, so the hidden cells are not in it. Must fill the clipboard programmatically.
3. This error occurs only if the first column is hidden, found solution.
4. Haha, I'm working for a German company and mix this all the time....... I did not see this anymore.
Changed!

I'm not sure with the additional warning, saving with hidden columns means deleting data now, but I think this is annoying on the long run. 

Thanks again!

**Comment by Isayso** [comment]
2nd pre-release available, not fully tested.

**Comment by dobbelina** [comment]
Hi again,
It's getting better and better!

Rename "Insert row" to "Paste row above", because "Insert row" only works if a row 
has been copied previously.
Even cooler would be if the renamed "Paste row above" is greyed out until "Copy row" is executed,
(memory contains a row)
but that is perhaps difficult to program? 

If "Show all columns" has been activated, and you save with "Ctrl + S", the
"spinning blue circle" appears, and no message.
Perhaps a pop-up could remedy that, which simply says "Playlist Saved" with an Ok button only.

The "Hidden Columns avaliable, continue?" message is unnecessary, because either
the hidden columns contain no data, or the user has made a deliberate choice 
not to keep them.
The same "Playlist Saved" with an Ok button only pop-up, as when all columns are visible
is better.

When trying the program out, i did a "#EXTM3U" search on pastebin, and it works
beatifully on most playlists.
There were a few with weird formatting, that it had difficulties to read,
but that is to be expected, as there is no formal specification for the M3U format.

I even found some nice new channels on pastebin while testing.
I noticed then how valuable it is that i can test play links directly from the program,
It's awesome.
Thanks for the great work so far!

**Comment by Isayso** [comment]
hi,
1. I made a method to check the clipboard content, so I can probably join both paste items to one. Or, I just grayed the Insert part out. What du you think?

2. I added a notification box, good idea

3. This warning is deleted.

Yes, the check link function is very helpful, I think. Its prolly not 100% right, because I have to set a timeout, if the link is very slow, this can lead to false response, but on very slow connection, you cannot see IPTV anyway. 
The corresponding function for the video playlist editor was much more difficult, because it has to check temporary available local network drives. 

Thanks!

**Comment by dobbelina** [comment]
Hi again Isayso,

1. The greyout you created on "Insert row" is perfect!
However, i would rename it to "Paste row", as "Insert row" is indicative of "Add row"
that we already have a button for.
"Paste row" is indicative of that you have copied something that you want to paste.

2. Your notification box "Playlist saved" is also perfect.
I love the fact that no interaction is needed like on a pop-up box with an ok button.

3. I noticed the "reload" icon, nice feature, so one can discard all changes without 
having to re-open the file, super!

4. Suggestion on the "Play with vlc" icon, I noticed that unless you have vlc installed
the icon doesn't show.
Wouldn't it be better if you keep the icon there anyway, and just throw another pop-up
to the user, encouraging them to install it if it's missing when pressed?
I think there's room for one "Play with kodi" button there as well.
Then you wouldn't need to have it in the context menu.
The "double-press" feature to start a player i would remove.
Reason being that you accidentally activate it when you want to edit a cell,
and that's kind of annoying.
Something like this perhaps, side by side, one for kodi & one for vlc:
<img width="242" alt="playbuttons" src="https://user-images.githubusercontent.com/46063764/66635524-b3664780-ec0f-11e9-93c7-e5bda8fdf4fd.png">


5. What does "Left string shift on paste" mean in the search settings?
With a default value of 0?

I really have to congratulate you on the progress so far, you are amazing!

**Comment by Isayso** [comment]
Many thanks!!!
1. the greyout feature was leading to a nice feature you will probably like. Direct import from clipboard, so copy a list from internet and import it. I'm sure there are still some errors in this function.
I changed "insert" to "paste".

2. I did similar for the vlc button, a red box with no vlc found, when not found ;)

3. I use reload quite often with the video playlist editor, when I want to move a hotkey grabbed link to an other list. 

4. I did some changes, hope you like it. 
I use the double click a lot, but see your complaint. I will make it switchable in the settings dialog, but that will take some time. 

5. I used a list with very annoying trailing characters like "[USA] ABC News", with left shift you can mark all cells and copy it to itself to delete the first 5 char "[USA]".  I found a bug for the last row, will fix it.  

Thank you for your ideas and testing!!!!!!!   I think they improved the prog a lot.

**Comment by dobbelina** [comment]
Hi again Isayso,

1. The "Import full list" from clipboard button is pure genius!
It streamlines the workflow to find new channels to 100%.
Import=>Linkcheck=>Remove duplicates=>Test play in vlc/Kodi = Brilliant!!

2. The addition of a kodi button, perfect.
Double click cell = vlc,Ctrl+P or Ctrl+double click cell = kodi, It's a keeper!
I don't think you need to re-program the double click feature to have the option 
to turn it off.

3. Minor issue,
I would remove the "vlc not installed" at start-up, perhaps the user don't want it installed,
And then it's only nagging.
It's perfectly sufficient to have it when clicking the vlc icon if not present.

4. I haven't been able to test the "Left shift string on paste" feature because i can't 
change 0 to anything else, it doesn't remember a change in the setting, it's always zero.
I have been thinking about an appropriate name change for that function also, so
that it becomes more intuitive.
"Remove characters left on paste,(+ shift)" perhaps?

**Comment by Isayso** [comment]
Hi dobbelina,
so, more progress

1. I think this function will not be 100% bug free, because I cannot test for valid clipboard content, but will work most of the time.  But no need to save as a file before. 

2. ok, I leave it as is

3. Found a bug, should work better now. 

4. The value is set to 0 when you open the settings window. The value is not saved. I added a tooltip to explain. And renamed it. 

5. found some strange resize behavior of the window, hopefully works now. Single column mode cannot be activated from fullscreen. 

Your input is very helpful! Program is much more useful now.

**Comment by dobbelina** [comment]
Hi again Isayso,

1. Aha!, now i understand the left character thing.
It works fine.
It's a good thing you made that change, as any new user seeing it for the
first time,(like me) would otherwise never figure it out, very good.

2. There is a weird issue with copy and paste cells when more than one cell
is blue-marked in a column.
If you copy 2 cells, selecting from top to bottom, (Downwards), and paste in place, both
cells get the same content.
However, if you select from bottom to top,(Upwards) and paste in place it works?
Similar odd behaviour happens when selecting more than one cell in a row.

3. If possible, I would prefer the "Paste row" function to paste it's content
to the same row as is selected, and not to the row above.
One could then use "Add row", and copy paste another row in it's place, (Like Excel).

I feel kind of unfair to come with all these suggestions, when you do
all the hard work, but i can promise you i appreciate it, you are truly one of a kind!
On the bright side you can be confident that this m3u editor is the best out there,
I am 100% sure.
Thanks for all your hard work.

**Comment by Isayso** [comment]
Hi dobbelina,
no, its not unfair, that step from a tool for yourself to every bodies tool is always hard work. So, please go on ;) 

2. Yes, its weird, the start point of paste is always the row with the triangle on the left.  When you drag a copy region the triangle follows to the last cell. Therefore the paste will start its work on the following cells. I have no good solution for this now. Its getting worse when you try to copy a region with  columns or not neighbored cells.  I've seen that Excel forbids coping not neighbored cells. 
Probably check if the triangle has moved between copy/paste and paste backwards. 

3. The "paste row" is more an "add row" now, right. I can add a real "paste row".

Because I have not found a useful editor for my needs, I started this project. For the video playlist I have not found any (?)  I'm sure that not many people will use the playlist function in Kodi, to use that feature with the remote control on a TV is such a pain. 

It's my pleasure to make it a useful program.

**Comment by dobbelina** [comment]
Hi again!

In only a few days you have transformed this application from a 
dull knife into a razor blade.
I'm very grateful.
As soon as you release a non pre-release version, I will update 
the kodi forum thread with your master piece, encouraging
people to download it.

**Comment by Isayso** [comment]
Hi,
I found some problems with copy/paste behavior, still working on it. 
What do you think to combine the copy cell (ctrl-c) and row (ctrl-r) to only one copy option? If you select the entire row on the left, rows will be copied, if you select cells only the cells will be copied. 

Before your help, it was a tool for my needs only, now its more and more a useful program.  
I'm very thankful for your input!

**Comment by dobbelina** [comment]
Hi,

Yes, good idea,
Something like this is the standard:
![edit-menu-copy-cut-paste](https://user-images.githubusercontent.com/46063764/66870672-400f5d80-efa2-11e9-81a3-d4b7b315018a.png)

**Comment by Isayso** [comment]
That was a lot of work, to do a combined copy I had to switch off the general copy/paste method and program my own. 
Should probably rename "Paste+Add" to "Insert" ......
  
Please try it out, the gray out logic was no fun too. 
Thanks again for your help!

**Comment by dobbelina** [comment]
Hi again!

Great work so far.
I would say it's working perfect, except when a single cell is copied.
The copy command doesn't respond if a single cell is only blue-marked.
It has to be clicked twice into edit-mode for the copy command to work.
Everything else, including the "Remove first character" function is flawless.

**Comment by Isayso** [comment]
I've seen this some minutes ago, too. Will look into it. When I drag a region its working (?) Hope its not a side effect of the general copy build in switch off. 
Cells are deleted now too, with "delete"

It was worth it!

**Comment by dobbelina** [comment]
Awesome!

Look forward to see next version.

**Comment by Isayso** [comment]
Yes, near by. Playing around with it to find something, looks good so far.  In single-column mode I kill the vlc now with a "program switch", remote control like. 

The "disable" logic in the context menu was hard to test. The c# code is literally switched off  and not accessible anymore with keyboard shortcuts when greyed out.

**Comment by dobbelina** [comment]
This program now officially gets my **AWESOME** stamp! 
https://forum.kodi.tv/showthread.php?tid=220615&pid=2893655#pid2893655

**Comment by Isayso** [comment]
Many thanks!
Without your input it wasn't possible. If you have more ideas let me know. 

thanks again.

**Comment by Isayso** [comment]
FYI: New version has a hotkey for import from clipboard, like the video Playlist editor. And a counter for the dead link check. Function can be canceled.

**Comment by dobbelina** [comment]
> 
> 
> FYI: New version has a hotkey for import from clipboard, like the video Playlist editor. And a counter for the dead link check. Function can be canceled.

Awesome,

Thank you!

**Comment by dobbelina** [comment]
Nice addition with the startup file. awesome!

**Comment by Isayso** [comment]
Thanks! If you have some ideas of missing functions, go ahead. 

The mru list works in the video editor, but needs lots of testing effort. Not sure if I will spend my time on this or focus on other features first.

**Comment by xjdacosta** [comment]
Hello,
very nice tool and interface and thank you to have written it.

I would like to ask/propose for some enhancements/correction:
1) when I select multiple rows in Column 'Title" and I do a "paste", the pattern in the buffer is applied only to the latest row and I would like to have it applied  to all the rows of the selected column
    --> that will allow to create/manage easily groups for tools like "Smart IPTV" (application for TV)
2) it would be nice if we could do a 'save as' for a part of the loaded file, by ex:
   I select some rows (the whole row) and if I do a "save as', at that time, it should do a 'save as' only of the selected rows and keep in the editor everything as it was before the 'save as' and not to change the name of the file (in the title bar) to the one we have terminated to 'save as'
3) enhance the Find feature
  when doing a search  through 'ctrl-f', it would be nice to have the possibility specify the action when the pattern is found:
 -> highlight the pattern simply (as currently)
 -> select the whole row where the pattern was found (that will allow to do a 'save as' of all these rows)
 -> select a specific column for all the rows having that pattern (that will allow to modify/paste a pattern in that column for all the rows selected:  typically, we search for the word france or hd that we know present in column "name2" and want to select the column title for all the rows having that pattern 
4) able to restrict the Find to the column or columns having a cell selected when we have pressed 'ctrl-f' : that will allow to do a find only on the column Name2 or in columns Name2 and Title if both have a cell selected.
5) a last one : enhance existing "load file":
    on top of the existing functionality (load a file and if later, we load another, the new replace the existing file);  
    allow to select/load multiple files (and merged them in the editor) or once a file was loaded, allow to load another file which will be appended to the one already loaded and not replacing it (could be achieved through the configure menu: by adding an option to change the behavioral of the load cmd : append or replace)

I do not know if you can or want and how it will be complex to implement all these enhancements but once, the tool will have them, it will become a real swiss knife to edit M3U files

Thanks

**Comment by Isayso** [comment]
Thanks,

just a quick answer, 

2: There is a workaround, select rows, Ctrl-c, Ctrl-n, ctrl-v. I think I can put this in the context menu. 

5: I already did a merge feature in the Playlist Editor (my other program) with d&d another file into it. Is this ok? Just drag adds the file, Ctrl-drag merging it. 

3: when you activate in options "Column search and select" "all cells" the whole row is selected and can be copied again with ctrl-c, ctrl-n. ctrl-v 

The other points, I must take a deeper look inside first to give you an answer.

have a nice day

**Comment by Isayso** [comment]
> 1. when I select multiple rows in Column 'Title" and I do a "paste", the pattern in the buffer is applied only to the latest row and I would like to have it applied  to all the rows of the selected column
>    --> that will allow to create/manage easily groups for tools like "Smart IPTV" (application for TV)

You want to have a fill option? Select a cell region and fill it with the same entry?


> 3. enhance the Find feature
>    when doing a search  through 'ctrl-f', it would be nice to have the possibility specify the action when the pattern is found:
>    -> highlight the pattern simply (as currently)
>    -> select the whole row where the pattern was found (that will allow to do a 'save as' of all these rows)

This is done with "all cell" option and copy the rows to new editor window opened with ctrl-n

>    -> select a specific column for all the rows having that pattern (that will allow to modify/paste a pattern in that column for all the rows selected:  typically, we search for the word france or hd that we know present in column "name2" and want to select the column title for all the rows having that pattern

Don't know exactly what you want to select. You want to find the column names which include  e.g. "HD" in their cells?

> 4. able to restrict the Find to the column or columns having a cell selected when we have pressed 'ctrl-f' : that will allow to do a find only on the column Name2 or in columns Name2 and Title if both have a cell selected.
 
Thats a logical find, right? So "HD" AND "tv", or so.

**Comment by Isayso** [comment]
> 5. a last one : enhance existing "load file":
>    on top of the existing functionality (load a file and if later, we load another, the new replace the existing file);
>    allow to select/load multiple files (and merged them in the editor) or once a file was loaded, allow to load another file which will be appended to the one already loaded and not replacing it (could be achieved through the configure menu: by adding an option to change the behavioral of the load cmd : append or replace)
> 

I already implemented it, drag an other file on the editor and it will be added at the end. 
click "search duplicates" with shift and the double rows will be deleted.

**Comment by xjdacosta** [comment]
hello,
item 5: feature exist already, as you have mentioned me, through the drag-and-drop feature. 

As it was not "working" trough the load menu, I made the assumption that it will be not working  through the drag-and-drop mechanism, that I have to admit I almost never use.

Thanks

**Comment by xjdacosta** [comment]
1) yes that will allow to create channels groups easily using the field group-title (column title) 

3) the functionality is there but there is a bug because if I do search for a pattern which is present (row is highlighted), I change in options menu, the value of "Column search and select" (still keeping the same pattern in the fields search), I press simply return on the search field after to have pressed ok in options menu: nothing is selected. I have to delete the pattern and type it again to obtain a selection of the new row corresponding to the section option.
 
I think it will better to dissociate the column options for the search and for the result selection (more intuitive for the users) and will add more flexibility if we could have 2 separated options:
- one to choose on which column we want to do the search
- a 2nd who says select what: whole row or a specific column 

4) yes

thanks

**Comment by Isayso** [comment]
I made a prototype with better find function with split between row/cell
and a fill from clipboard context menu entry.
Please try it and tell me what you think.

You find the file [TestVersion](test/PlaylistEditorIPTV.exe)
test/PlaylistEditorIPTV.exe

**Comment by Isayso** [comment]
I released 1.4.7 with changes you will hopefully like. 
Thanks for your ideas!

**Comment by Isayso** [comment]
Thanks for your ideas!
